### PR TITLE
Replace AWS sigv2 with sigv4

### PR DESCRIPTION
**Pull request by lbakerchef**
This PR is part of a broader effort to replace AWS sigv2 with sigv4, encompassing modifications to oc_erchef, bookshelf, mini_s3, erlcloud, pedant clients, and other systems. mini_s3 was essentially turned both into a wrapper for erlcloud, and an adapter to handle impedence mismatches.  The expiration windows feature was reengineered.  New unit tests exercising various aspects of newly-added functionality were also added.

See:
https://github.com/chef/chef-server/issues/1911
https://github.com/chef/chef-server/pull/1993
https://github.com/chef/erlcloud/pull/2

- [x] Green verify pipeline
- [x] Green omnibus/adhoc pipeline
- [x] Green integration test pipeline
- [x] Green dialyzer pass

**Expiration Windows Change**

I investigated simply adapting the old sigv2 expiration window implementation to sigv4, but it didn't seem straightforward to do.  

**Requirements**

- The old sigv2 implementation outputs a **future** expiration time, in a **particular time format**. The sigv4 implementation requires output of 1) a time in the **past**, in a **different time format** than sigv2 implementation, and 2) a lifetime in seconds.

- Calculation of 'present time' internally within the two implementations seems to require different time formats.

**Day Boundary**

There seems to be some kind of correction for a 'day boundary' being crossed in the old sigv2 implementation.  I don't fully understand the issue, but at any rate, AWS sigv4 merely requires a time in the past, and a lifetime in seconds until expiration.  I would be surprised if the S3 service couldn't handle a timestamp of 11:55pm with a lifetime in seconds that caused the link to expire after midnight.

**Rationale for Reimplementation**

It's an open question as to whether the different time formats are convertible between the two, but at the time I did not want to spent an inordinate amount of developer time figuring this out, especially since both implementations (sigv2 and sigv4) are short.  Also, if the time formats are convertible, the amount of adaptive code necessary to convert between the two formats in an 'adapt the sigv2 version to sigv4' scenario would likely make the resulting 'sigv2 to sigv4 adapted' implementation longer than the straight sigv4 re-implementation.

Having said that, if anyone 1) sees a problem with the new implementation, or 2) has an idea for simply converting or adapting the old implementation, I would be happy to do it.  Since the old version has withstood the test of time, it might be a desirable way to go.

**Comparison of Output**

I compared some output of the two expiry windows implementations.  This study was by no means exhaustive, but the implementations seem to function similarly for a few inputs I gave.

The old implementation is labeled _expiration_time_.  The new implementation is labeled _make_expire_win_.

Here are 5 expire windows for both implementations, 1 window for each simultaneously created per second, with TTL=0, WinSiz=1. Notice that the last two digits for both implementations tick up 1 second for each expire window. I did not synchronize them, but they acted synchronized and seemed to be doing the same thing:

```
f(), TTL=0, WinSiz=1, [{io:format("~nexpiration_time: ~p~nmake_expire_win: ~p", [mini_s3:expiration_time({TTL, WinSiz}), mini_s3:make_expire_win(TTL, WinSiz)]), timer:sleep(1000)} || _ <- lists:duplicate(5,0)].

expiration_time: 1605217778
make_expire_win: {"20201112T214937Z",1}
expiration_time: 1605217779
make_expire_win: {"20201112T214938Z",1}
expiration_time: 1605217780
make_expire_win: {"20201112T214939Z",1}
expiration_time: 1605217781
make_expire_win: {"20201112T214940Z",1}
expiration_time: 1605217782
make_expire_win: {"20201112T214941Z",1}
```

Here are 5 expire windows for both implementations, 1 window for each simultaneously created per second, with TTL=0, WinSiz=5. Notice that the last two digits for both implementations stay the same for the first two windows,
then both tick up 5 seconds for the third window and remain the same (apparently, we crossed a window here). Again, i did not synchronize these, but they acted synchronized and seem to be doing the same thing:

```
f(), TTL=0, WinSiz=5, [{io:format("~nexpiration_time: ~p~nmake_expire_win: ~p", [mini_s3:expiration_time({TTL, WinSiz}), mini_s3:make_expire_win(TTL, WinSiz)]), timer:sleep(1000)} || _ <- lists:duplicate(5,0)].

expiration_time: 1605217845
make_expire_win: {"20201112T215040Z",5}
expiration_time: 1605217845
make_expire_win: {"20201112T215040Z",5}
expiration_time: 1605217850
make_expire_win: {"20201112T215045Z",5}
expiration_time: 1605217850
make_expire_win: {"20201112T215045Z",5}
expiration_time: 1605217850
make_expire_win: {"20201112T215045Z",5}
```

Finally, here are 5 expire windows for both implementations, 1 window for each simultaneously created per second, with TTL=5, WinSiz=1. Notice that the last two digits for both implementations tick up 1 second for each expire window. I did not synchronize them, but they acted synchronized and seem to be doing the same thing:

```
f(), TTL=5, WinSiz=1, [{io:format("~nexpiration_time: ~p~nmake_expire_win: ~p", [mini_s3:expiration_time({TTL, WinSiz}), mini_s3:make_expire_win(TTL, WinSiz)]), timer:sleep(1000)} || _ <- lists:duplicate(5,0)].

expiration_time: 1605217964
make_expire_win: {"20201112T215238Z",5}
expiration_time: 1605217965
make_expire_win: {"20201112T215239Z",5}
expiration_time: 1605217966
make_expire_win: {"20201112T215240Z",5}
expiration_time: 1605217967
make_expire_win: {"20201112T215241Z",5}
expiration_time: 1605217968
make_expire_win: {"20201112T215242Z",5}
```

See `test/mini_s3_tests.erl` for testing of expiration window properties which should hold.

### CODE REVIEW

File: src/mini_s3.erl
Comment (PrajaktaPurohit) [inline src/mini_s3.erl]: I am not sure this comment adds much.

File: src/erlcloud_aws.hrl
Comment (stevendanna) [inline src/erlcloud_aws.hrl:1]: Let's remove this vendored file or document exactly what problem we were hitting that led us to add it. Further if we do add it here, we'll need to include the license.

File: rebar.config
Comment (stevendanna) [inline rebar.config:6]: Is ibrowse needed any longer? I don't think it should be.

File: src/mini_s3.erl
Comment (stevendanna) [inline src/mini_s3.erl:72]: Can we clean up this commented code? What do we need to figure out here?

File: src/mini_s3.erl
Comment (stevendanna) [inline src/mini_s3.erl:142]: It might be nice to pull this URI parsing code out into its own function.

File: test/mini_s3_tests.erl
Comment (stevendanna) [inline test/mini_s3_tests.erl:57]: I'm worried that we are removing these tests as I don't see what other tests replace them in terms of testing the main invariants of our cachable URLs

File: src/mini_s3.erl
Comment (stevendanna) [inline src/mini_s3.erl:344]: In the previous code we were concerned about the day boundary? Is that a concern anymore? I would have thought it is more of a concern because of how SigV4 handles the credential context.

File: src/erlcloud_aws.hrl
Comment (lbakerchef) [reply src/erlcloud_aws.hrl]: File deleted.
https://buildkite.com/chef-oss/chef-mini-s3-master-verify/builds/5
https://buildkite.com/chef/chef-chef-server-master-verify/builds/2425#16139667-3e0b-4d14-b425-a855b4f86e48
https://buildkite.com/chef/chef-chef-server-master-omnibus-adhoc/builds/1659#1e73e43d-30bc-4cad-ac2f-f8627dcd5fb8

File: src/mini_s3.erl
Comment (stevendanna) [inline src/mini_s3.erl:213]: Dropping SSL Options is going to be a problem for A2-based Chef Server's as we need to be able to pass the SSL Options to talk to bookshelf with the right TLS configuration.

File: src/mini_s3.erl
Comment (lbakerchef) [reply src/mini_s3.erl]: That comment was very helpful to explain to myself what I had done when I had to "re- figure out" what I had done :)  Do you mind if we keep it, for my own future sanity?

File: src/mini_s3.erl
Comment (lbakerchef) [reply src/mini_s3.erl:72]: Fixed.

File: src/mini_s3.erl
Comment (lbakerchef) [reply src/mini_s3.erl]: Done.
https://buildkite.com/chef-oss/chef-mini-s3-master-verify/builds/9
https://buildkite.com/chef/chef-chef-server-master-verify/builds/2437#2f07bc74-f096-43b5-8848-33dcbf628443
https://buildkite.com/chef/chef-chef-server-master-omnibus-adhoc/builds/1667#7b94b4e5-b553-4482-b187-09fbd6fc3a5c

File: rebar.config
Comment (lbakerchef) [reply rebar.config]: ibrowse deleted.
https://buildkite.com/chef-oss/chef-mini-s3-master-verify/builds/8
https://buildkite.com/chef/chef-chef-server-master-verify/builds/2436#b9944907-8fe2-4dd2-94c3-9e3dbb97278c
https://buildkite.com/chef/chef-chef-server-master-omnibus-adhoc/builds/1666#a062a6e9-23e2-4239-a804-e46ab54c2872

File: src/mini_s3.erl
Comment (lbakerchef) [reply src/mini_s3.erl:213]: i'll have to investigate this.

Comment (lbakerchef) [reply src/mini_s3.erl:213]: @PrajaktaPurohit and I did a 'mini triage' on this.  She seems to think we can kick the can down the road for a bit, and revisit later (possibly even after a merge).  Do you concur?

File: src/mini_s3.erl
Comment (lbakerchef) [reply src/mini_s3.erl:344]: I tried to look into this somewhat.  I wish I understood the issue better (perhaps you can explain it to me).  And I wish I understood pre-existing code better.  What I can say is, my code does all calculations on 'seconds' (seconds since whenever timekeeping started haha) and has absolutely no concept about days or day boundaries.  At the end, it converts these seconds into a date/time.  So does a lack of a concept of days and day boundaries mean my code has a problem?  Or does it mean the opposite - that there can't be a problem because it has no concept of such things?  I don't know.

Perhaps we can look into this together.  I can explain to you how my code works, and draw up diagrams as necessary.  If you know how the previous code works, we can determine if there is an issue.  Perhaps you can also give me a 'day boundary' to run through my code, and we can see if it comes up with the correct calculation or not.

Comment (lbakerchef) [reply src/mini_s3.erl:344]: One thing is, sigv4 presigned URLs cannot be valid for more than 1 week.  But my code probably happily allows you to create presigned URLs that are greater than one week.  I don't know if this is a problem.  My assumption was that we should simply not ask for presigned URLs which are valid for more than one week, i.e. don't specify a TTL or an expiration window of a week or greater.

Comment (lbakerchef) [reply src/mini_s3.erl:344]: Thinking about the above more, it seems at the very least we could have a week boundary issue.  It might be possible to specify inputs that 'accidentally' give expirations greater than 1 week.  I can draw this out for you to see if you concur.  Perhaps I can add code to simply curb or squelch any expiration going over a week to a week.

Comment (lbakerchef) [reply src/mini_s3.erl:344]: Added squelching of output URL lifetimes to 1 week.

Comment (lbakerchef) [reply src/mini_s3.erl:344]:
```
% specified TTL of greater than 1 week (604800 seconds)
% output lifetime is squelched to 1 week

> mini_s3:make_expire_win(999999,1).
{"20200916T215530Z",604800}

% specified expiration window of greater than 1 week (604800 seconds)
% output lifetime is squelched to 1 week

9> mini_s3:make_expire_win(1,999999).
{"20200910T055033Z",604800}
```

Comment (lbakerchef) [reply src/mini_s3.erl:350]: Also, just checked.  It seems I reject presigned URLs with expirations greater than 1 week in bookshelf authentication (bksw_sec.erl).

File: src/mini_s3.erl
Comment (stevendanna) [inline src/mini_s3.erl:149]: Have we looked into using this: http://erlang.org/doc//man/uri_string.html I wonder if it might get us most of what we want? But, you've written enough tests that this isn't my biggest concern, so no worries if not.

**Review by stevendanna — COMMENTED** [review]
I think the main things we want to do here are:

- Investigate SSL options for erlcloud so it works in Automate
- See if we can get the same test cases for the expiry window working, or at least document why they have changed and how.

**Comment by lbakerchef** [comment]
Blog on expiry windows, courtesy of @PrajaktaPurohit :
https://blog.chef.io/cookbook-caching/

File: src/mini_s3.erl
Comment (lbakerchef) [reply src/mini_s3.erl:138]: It seems the first case must be matched by Host0 which is passed-in.  The last case is more general.

Comment (stevendanna) [reply src/mini_s3.erl:138]: Ah, missed that was already a bound variable

Comment (lbakerchef) [reply src/mini_s3.erl:138]: Easy to miss.

File: rebar.config
Comment (lbakerchef) [reply rebar.config:5]: Done.

File: src/mini_s3.erl
Comment (lbakerchef) [reply src/mini_s3.erl:149]: After reading your comment, I did quite a bit of work converting over to `uri_string`.  The issue was, it does not correctly parse `"host:port"`, whether host is ipv4, or ipv6.  I was able to code up a correction in the ipv4 case without adding too much code or complexity.  But if we correct for the ipv6 case, the resulting amount of code/complexity will be more than if we simply go with what we have.  Your call, and I'll do whatever you recommend.

File: rebar.config
Comment (lbakerchef) [reply rebar.config:5]: Done.

File: src/mini_s3.erl
Comment (lbakerchef) [reply src/mini_s3.erl:149]: EDIT:  "Hold the phone, Chuck."  I may have an acceptable solution.

Comment (lbakerchef) [reply src/mini_s3.erl]: Done.

File: rebar.config
Comment (PrajaktaPurohit) [inline rebar.config:15]: We should add the warnings as error back in.

Comment (lbakerchef) [reply rebar.config]: Done.

File: src/mini_s3.erl
Comment (PrajaktaPurohit) [inline src/mini_s3.erl:140]: change x to missing_scheme maybe?

File: src/mini_s3.erl
Comment (PrajaktaPurohit) [reply src/mini_s3.erl:200]: I think since it is 2 months past that discussion - now is a good time to revisit this issue :) We should probably create a ticket if we haven't already for this.

File: src/mini_s3.erl
Comment (lbakerchef) [reply src/mini_s3.erl]: Done.

File: src/mini_s3.erl
Comment (lbakerchef) [reply src/mini_s3.erl:346]: See some analysis further down on this.

Comment (lbakerchef) [reply src/mini_s3.erl:346]: See some analysis further down on this.

Comment (PrajaktaPurohit) [reply src/mini_s3.erl:348]: Looks like the day boundary that was being handles was for the purpose of simplifying the expiration times for the user, it seems to not have any effect on caching. Looking at https://docs.aws.amazon.com/general/latest/gr/sigv4-date-handling.html, It seems like as long as the x-amz-date header value (which is the time the request was made)  is the same as the date component in the credential scope, the authentication should be successful. So keeping the date boundary would be good to have to not break user expectations.

Comment (PrajaktaPurohit) [reply src/mini_s3.erl:348]: That being said, @lbakerchef the Lifetime returned in the make_expire_win function did not match what I expected to see in the case where TTL = 60 mins and Interval = 15 mins. Would like to understand if I am missing something here.

Comment (lbakerchef) [reply src/mini_s3.erl:348]: @PrajaktaPurohit is this conversation resolved since you added-in the old expiry windows tests?  Or should I leave this open?

File: src/mini_s3.erl
Comment (lbakerchef) [reply src/mini_s3.erl:215]: Ticket created.

File: src/mini_s3.erl
Comment (lbakerchef) [reply src/mini_s3.erl:348]: Reminder: Add @PrajaktaPurohit 's contribution to this branch's code.

**Comment by lbakerchef** [comment]
Test @PrajaktaPurohit 's expiration windows modifications:

```
lbakers-MacBook-Pro:mini_s3 lbaker$ ./rebar3 shell
===> Verifying dependencies...
===> Compiling mini_s3
Erlang/OTP 23 [erts-11.1.8] [source] [64-bit] [smp:12:12] [ds:12:12:10] [async-threads:1] [hipe] [dtrace]

Eshell V11.1.8  (abort with ^G)
7> Config = mini_s3:new("key", "secret", "http://test.com").
{aws_config,"autoscaling.amazonaws.com","ec2.amazonaws.com",
            "iam.amazonaws.com","sts.amazonaws.com","http://",
            "test.com",80,false,2,path,true,"sdb.amazonaws.com",
            "elasticloadbalancing.amazonaws.com",
            "rds.us-east-1.amazonaws.com",
            "email.us-east-1.amazonaws.com","queue.amazonaws.com",
            undefined,undefined,"https://",
            "elasticmapreduce.us-east-1.amazonaws.com",undefined,
            "https://","sns.amazonaws.com",undefined,
            "mechanicalturk.amazonaws.com","monitoring.amazonaws.com",
            undefined,undefined,...}
8> TTL = 300.
300
9> ExpireWin = 300.
300
10> % time = 2:16pm
10> rp(mini_s3:s3_url(get, "my-bucket", "my-key", {TTL, ExpireWin}, [], Config)).
<<"http://test.com/my-bucket/my-key?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=key%2F20210415%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20210415T191500Z&X-Amz-Expires=600&X-Amz-SignedHeaders=host&X-Amz-Signature=650d7b6f1678239c684a6971844e9679a45718365c76cc2bbf0c7bd881a6adb5">>
ok
11> % time = 2:17pm
11> rp(mini_s3:s3_url(get, "my-bucket", "my-key", {TTL, ExpireWin}, [], Config)).
<<"http://test.com/my-bucket/my-key?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=key%2F20210415%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20210415T191500Z&X-Amz-Expires=600&X-Amz-SignedHeaders=host&X-Amz-Signature=650d7b6f1678239c684a6971844e9679a45718365c76cc2bbf0c7bd881a6adb5">>
ok
12> % time = 2:18pm
12> rp(mini_s3:s3_url(get, "my-bucket", "my-key", {TTL, ExpireWin}, [], Config)).
<<"http://test.com/my-bucket/my-key?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=key%2F20210415%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20210415T191500Z&X-Amz-Expires=600&X-Amz-SignedHeaders=host&X-Amz-Signature=650d7b6f1678239c684a6971844e9679a45718365c76cc2bbf0c7bd881a6adb5">>
ok
13> % time = 2:19pm
13> rp(mini_s3:s3_url(get, "my-bucket", "my-key", {TTL, ExpireWin}, [], Config)).
<<"http://test.com/my-bucket/my-key?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=key%2F20210415%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20210415T191500Z&X-Amz-Expires=600&X-Amz-SignedHeaders=host&X-Amz-Signature=650d7b6f1678239c684a6971844e9679a45718365c76cc2bbf0c7bd881a6adb5">>
ok
14> % time = 2:20pm
14> rp(mini_s3:s3_url(get, "my-bucket", "my-key", {TTL, ExpireWin}, [], Config)).
<<"http://test.com/my-bucket/my-key?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=key%2F20210415%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20210415T191500Z&X-Amz-Expires=600&X-Amz-SignedHeaders=host&X-Amz-Signature=650d7b6f1678239c684a6971844e9679a45718365c76cc2bbf0c7bd881a6adb5">>
ok
15> % time = 2:21pm
15> rp(mini_s3:s3_url(get, "my-bucket", "my-key", {TTL, ExpireWin}, [], Config)).
<<"http://test.com/my-bucket/my-key?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=key%2F20210415%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20210415T191500Z&X-Amz-Expires=600&X-Amz-SignedHeaders=host&X-Amz-Signature=650d7b6f1678239c684a6971844e9679a45718365c76cc2bbf0c7bd881a6adb5">>
ok
16> % time = 2:22pm
16> rp(mini_s3:s3_url(get, "my-bucket", "my-key", {TTL, ExpireWin}, [], Config)).
<<"http://test.com/my-bucket/my-key?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=key%2F20210415%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20210415T192000Z&X-Amz-Expires=600&X-Amz-SignedHeaders=host&X-Amz-Signature=00308fdaf1702218ad8d5fe9bc430236d23acde8f6f689b12f53bca72869a3d0">>
ok
```

**Review by vinay-satish — APPROVED** [review]
Code looks good to me, could you please add the links to the build kite pipelines.

**Comment by lbakerchef** [comment]
> Code looks good to me, could you please add the links to the build kite pipelines.

@vinay-satish See PR here for adhocs/umbrellas:
https://github.com/chef/chef-server/pull/2097

**Review by marcparadise — APPROVED** [review]

**Review by PrajaktaPurohit — APPROVED** [review]

**Pull request merged**
